clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -304,126 +317,79 @@ static Expected<StringRef> linkDeviceLibFiles(StringRef InputFile,
-  getSPIRVTransOpts(Args, CmdArgs, Triple);
-  StringRef LLVMToSPIRVOptions;
-  if (Arg *A = Args.getLastArg(OPT_llvm_spirv_options_EQ))
-    LLVMToSPIRVOptions = A->getValue();
-  LLVMToSPIRVOptions.split(CmdArgs, " ", /* MaxSplit = */ -1,
-                           /* KeepEmpty = */ false);
-  CmdArgs.append({"-o", OutputFile});
-  CmdArgs.push_back(File);
-  if (Error Err = executeCommands(*LLVMToSPIRVProg, CmdArgs))
-    return std::move(Err);
-
-  if (!SPIRVDumpDir.empty()) {
-    std::error_code EC =
-        llvm::sys::fs::create_directory(SPIRVDumpDir, /*IgnoreExisting*/ true);
-    if (EC)
-      return createStringError(
-          EC,
-          formatv("failed to create dump directory. path: {0}, error_code: {1}",
-                  SPIRVDumpDir, EC.value()));
-
-    StringRef Path = OutputFile;
-    StringRef Filename = llvm::sys::path::filename(Path);
-    SmallString<128> CopyPath = SPIRVDumpDir;
-    CopyPath.append(Filename);
-    EC = llvm::sys::fs::copy_file(Path, CopyPath);
-    if (EC)
-      return createStringError(
-          EC,
-          formatv(
-              "failed to copy file. original: {0}, copy: {1}, error_code: {2}",
-              Path, CopyPath, EC.value()));
+/// be parsed to generate options required to be passed into the backend.
+static Expected<StringRef> runSPIRVCodeGen(StringRef File,
+                                           const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("SPIR-V code generation");
+  if (Verbose || DryRun) {
+    errs() << formatv("LLVM-SPIRV-Backend: input: {0}, output: {1}\n", File,
+                      OutputFile);
+    if (DryRun)
+      return OutputFile;
   }
 
+  // SPIR-V-specific target initialization.
+  InitializeSPIRVTarget();
+
+  SMDiagnostic Err;
+  LLVMContext C;
+  std::unique_ptr<Module> M = parseIRFile(File, Err, C);
+  if (!M)
+    return createStringError(inconvertibleErrorCode(), Err.getMessage());
+
+  static const std::string DefaultTriple = "spirv64v1.6-unknown-unknown";
+
+  // Correct the Triple value if needed
+  // TODO: Remove this correction once we start using spirv64/spirv32 triples
+  // everywhere.
+  Triple TargetTriple(M->getTargetTriple());
+  if (TargetTriple.isSPIR()) {
+    TargetTriple.setArch(TargetTriple.getArch() == Triple::spir64
+                             ? Triple::spirv64
+                             : Triple::spirv32,
+                         TargetTriple.getSubArch());
+    M->setTargetTriple(TargetTriple);
+    // We need to reset Data Layout to conform with the TargetMachine
+    M->setDataLayout("");
+  }
+  if (TargetTriple.getTriple().empty())
+    TargetTriple.setTriple(DefaultTriple);
+  TargetTriple.setArch(TargetTriple.getArch(), Triple::SPIRVSubArch_v16);
+  M->setTargetTriple(TargetTriple);
+
+  std::string Msg;
+  const Target *T = TargetRegistry::lookupTarget(M->getTargetTriple(), Msg);
+  if (!T)
+    return createStringError(Msg + ": " + M->getTargetTriple().str());
```

Comment:
@VyacheslavLevytskyy, do we need any of this code? What will happen if SPIR-V backend compile LLVM module with spir (not spirv) target triple?

Reply:
Hi @bader 

This is a good catch. I do not think we need this check here. We can work on the premise that the SYCL compilation flow will use spirv32/spirv64 as triples for JIT compilations. 

I tried with spir64 triple and I get the following error:
TargetTriple = spir64-unknown-unknown
clang-sycl-linker: error: No available targets are compatible with triple "spir64-unknown-unknown": spir64-unknown-unknown

I will remove the checks and update the tests to create fat objects using SPIRV triples.

Thanks